Integrates exploration space and experiment bundle generation from DSL

jeppe-style opened the pull request:
Fixes issues for test-manager: 
- [X] https://github.com/benchflow/benchflow/issues/370 
- [X] https://github.com/benchflow/benchflow/issues/257

Partially fixes for test-manager: 
- [X] for including `one-at-a-time` and `random_validation` selection strategies https://github.com/benchflow/benchflow/issues/287
- [X] this PR also adds validation strategy for the complete/exhaustive exploration space (https://github.com/benchflow/benchflow/issues/287). 

Open issues after this:
- https://github.com/benchflow/benchflow/issues/391

jeppe-style committed:
Updates DSL library version
Updates tests accordingly.

jeppe-style committed:
Updates state transitions for running state according to updated state chart

jeppe-style committed:
Includes updated DSL version

jeppe-style committed:
Validates one-at-a-time selection strategy for exploration without regression.

jeppe-style committed:
Adds verification of expected order of experiments

jeppe-style committed:
Moves validation of complete exploration into validation strategy

jeppe-style committed:
Adds random breakdown selection strategy
Reverts to using DSL currently in devel branch

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/scheduler/TestTaskScheduler.java:
```diff
@@ -233,12 +240,48 @@ private void addStoredKnowledge(String testID) {
 
     // TODO - should go into a stateless queue (so that we can recover)
-    Future<?> future = taskExecutorService.submit(task);
+    Future<Boolean> future = taskExecutorService.submit(task);
```
Determining if we require a prediction model should happen in the `scheduler`

VincenzoFerme replied:
Done in https://github.com/benchflow/benchflow/pull/394/commits/ab816423446799a9f6e388d52f85c15bef9a4c58

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/scheduler/TestTaskScheduler.java:
```diff
@@ -383,34 +439,21 @@ private synchronized void handleExperimentResult(String testID) {
 
     // TODO - should go into a stateless queue (so that we can recover)
-    Future<HandleExperimentResultTask.Result> future = taskExecutorService.submit(task);
+    Future<Boolean> future = taskExecutorService.submit(task);
```
You should determine if you need a regression model in the `scheduler`

VincenzoFerme replied:
Done in https://github.com/benchflow/benchflow/pull/394/commits/ab816423446799a9f6e388d52f85c15bef9a4c58

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/tasks/running/ValidateTerminationCriteria.java:
```diff
@@ -27,13 +28,31 @@ public ValidateTerminationCriteria(String testID) {
+
+  }
+
+  public enum TerminationCriteriaResult {
```
Come up with better naming for these statuses

VincenzoFerme replied:
Done in https://github.com/benchflow/benchflow/pull/394/commits/51f1a8aefa04cacbcc20128cb82fe45829144d93

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/services/internal/dao/ExplorationModelDAO.java:
```diff
@@ -26,62 +30,182 @@ public ExplorationModelDAO(MongoClient mongoClient, BenchFlowTestModelDAO testMo
+
+  }
+
+  public synchronized List<Integer> getExplorationPointIndices(String testID)
```
Change the name to `getExecutedExplorationPointIndices`

VincenzoFerme replied:
Done in https://github.com/benchflow/benchflow/pull/394/commits/31ad2fa5fce3770738e2a0be4eeef66410c944d9

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/strategy/selection/OneAtATimeSelectionStrategy.java:
```diff
@@ -0,0 +1,106 @@
+            explorationPointIndices.stream().max(Integer::compareTo).orElse(0) + 1;
+      }
+
+      JavaCompatExplorationSpace explorationSpace = explorationModelDAO.getExplorationSpace(testID);
```
Move to the `SelectionStrategy` and have it as an abstract class

VincenzoFerme replied:
Carefully separate what is about selecting a point in the space and what is about generating the bundle, so that you enforce the separation of concerns in the implementer

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/tasks/running/DetermineExecuteExperimentsTask.java:
```diff
@@ -55,19 +54,23 @@ public void run() {
+      experimentModelDAO.setExplorationSpaceIndex(experimentID,
+          selectedExperimentBundle.getExplorationSpaceIndex());
 
       // save PE defintion to minio
```
==> `BE`

VincenzoFerme replied:
Done in https://github.com/benchflow/benchflow/pull/394/commits/3169f96597122c1c799e101f6ba8b10b180ebafd

VincenzoFerme commented on benchflow-test-manager/application/src/test/java/cloud/benchflow/testmanager/strategy/selection/OneAtATimeSelectionStrategyTest.java:
```diff
@@ -53,33 +53,34 @@ public void selectNextExperiment() throws Exception {
-
-    Assert.assertNotNull(experimentYaml);
-    Assert.assertTrue(experimentYaml.contains("users: " + expectedNumUsers));
+    //    List<Integer> selectionStrategy = StartTask.generateExplorationSpace(test);
```
Update to the new code, or describe in a comment why it is commented out

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/tasks/running/DetermineExecuteExperimentsTask.java:
```diff
@@ -65,12 +65,10 @@ public void run() {
+          selectedExperimentBundle.getExplorationSpaceIndex());
 
-      // save PE defintion to minio
+      // save PE definition to minio
```
==> `BE`

VincenzoFerme replied:
Done in https://github.com/benchflow/benchflow/pull/394/commits/3169f96597122c1c799e101f6ba8b10b180ebafd

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/strategy/selection/RandomBreakdownSelectionStrategy.java:
```diff
@@ -0,0 +1,103 @@
+
+  private int getNextExplorationPoint(int upperBound, List<Integer> explorationPointIndices) {
+
+    Random random = new Random();
```
Make sure that this random, cannot generate an ordered sequence. For example selecting a seed that guarantees that the first value that is picked is not the first value in the `explorationPointIndices` or by checking sizes.

jeppe-style replied:
@VincenzoFerme I did not find anything about this. I think also it is not possible to have such guarantees. Only way (I think) would be to generate the whole list and the shuffle it, and then checking that it is not sorted. Also I'm not sure that it is good to avoid having the same order as `one-at-a-time`. If that is what the random generator comes up with, that is also random.

VincenzoFerme replied:
I agree with you, so please update the test so that it gets executed a given amount of time, and checks that the order you get after each execution is different. A good idea is to use statistics to test it: https://stackoverflow.com/a/2130691

jeppe-style replied:
I can run the test 3 times to check that each case is different. I think that is sufficient because we should be able to rely on that the number generation from Java is random.

jeppe-style committed:
Refactors if regression model check from AddStoredKnowledgeTask to scheduler

jeppe-style committed:
Refactors if regression model check from HandleExperimentResultTask to scheduler

jeppe-style committed:
Improves naming for TerminationCriteriaResult

jeppe-style committed:
Renames method in ExplorationModelDAO

jeppe-style committed:
Refactors names according to thesis
archive => bundle, performance experiment manager => experiment manager, performance test manager => test manager

VincenzoFerme commented:
Still unresolved:

- [X] https://github.com/benchflow/benchflow/pull/394#discussion_r122488467 - Done in https://github.com/benchflow/benchflow/pull/394/commits/f59dd7d61db81ebecc7f6c4fa064cf08449ec451
- [x] https://github.com/benchflow/benchflow/pull/394#discussion_r122491391 - Done in https://github.com/benchflow/benchflow/pull/394/commits/d8e8684ac82048a64de99d3e31d9a55976287c93
- [x] https://github.com/benchflow/benchflow/pull/394#discussion_r122494232 - Done in https://github.com/benchflow/benchflow/pull/394/commits/2caec4cf8b4ad7895c032434e7a63a20b238107f
- [x] https://www.codacy.com/app/BenchFlow/benchflow/pullRequest?prid=690563

jeppe-style committed:
Refactors common code into SelectionStrategy
Makes SelectionStrategy abstract to incorporate common code.

jeppe-style committed:
Adds comment to clarify why OneAtATimeSelectionStrategyTest is commented

jeppe-style committed:
Fixes codacy warnings

jeppe-style committed:
Comments out a test implemented in following PR

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/strategy/selection/SelectionStrategy.java:
```diff
@@ -1,17 +1,100 @@
   }
 
-  SelectedExperimentBundle selectNextExperiment(String testID);
+  public class SelectedExperimentBundle {
```
This can probably be `private` or `protected`.

jeppe-style replied:
It is used by other classes outside the package. We can also separate it into its own class if you think that is better. It is now there just because it is related to the `SelectionStrategy`

jeppe-style committed:
Adds test to check random-breakdown selection strategy
Removes previous test that was not correct and could fail.

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/strategy/selection/SelectionStrategy.java:
```diff
@@ -1,17 +1,100 @@
+
+      // get the experiment bundle for the given point
+      // <ExperimentDefinition, DeploymentDescriptor>
+      Tuple2<String, String> experimentBundle =
```
Isn't it better to have a Java object to store the two artefacts of the bundle?

jeppe-style replied:
I don't understand fully the issue.

VincenzoFerme replied:
It is ok, you encapsulate the objects in `SelectedExperimentBundle` later on.

VincenzoFerme commented on benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/strategy/selection/SelectionStrategy.java:
```diff
@@ -1,17 +1,100 @@
   }
 
-  SelectedExperimentBundle selectNextExperiment(String testID);
+  public class SelectedExperimentBundle {
```
It is instantiated by other class outside the package? It is fine it is in the `SelectionStrategy` class, and I think the only class that instantiate it is `SelectionStrategy`, so it can be private.
